### PR TITLE
Zdc unpacker digi reco fix81 x

### PR DESCRIPTION
**Pull request by cferraio**
Updated PR #13881 with ALL of the files instead of just some of them :-/.

I've updated the zdc unpacker, which was implemented in 75X for the HI run, but not in subsequent releases. Additionally, I've updated digi->reco functionality so that reco level ZDC objects are generated.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by cferraio** [comment]
Updated code with a minor change. Looks like the error occurred because the code worked for data but not MC. I tested using "runTheMatrix.py -s --useInput all &", which returned no errors. Thanks.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cferraio 
I see from the first few files that changes made here are pretty close to what was put to 75X.
 https://github.com/cms-sw/cmssw/pull/12497#issuecomment-158431882
The main reason this was not propagated to 76X/8X was that we expected a more appropriate update (separate ZDC raw2digi, something that covers both run1 and run2 configurations).

Please clarify what's in this PR.

File: RecoLocalCalo/Configuration/python/RecoLocalCalo_EventContent_cff.py [hunk 1]
```diff
@@ -32,6 +32,7 @@
                                            'keep HFRecHitsSorted_hfrecoMB_*_*',
                                            #'keep ZDCDataFramesSorted_*Digis_*_*',
                                            'keep ZDCDataFramesSorted_hcalDigis_*_*',
+                                           'keep ZDCDataFramesSorted_castorDigis_*_*',
```
Comment (slava77) [inline RecoLocalCalo/Configuration/python/RecoLocalCalo_EventContent_cff.py]: I think that the same change should be added to RecoLocalCalo_EventContentCosmics_cff.py as well

File: EventFilter/CastorRawToDigi/plugins/CastorRawToDigi.cc [hunk 2]
```diff
@@ -72,37 +82,93 @@ void CastorRawToDigi::produce(edm::Event& e, const edm::EventSetup& es)
...
+	  }
       }
-    }
-  }
+
+    if (*i==693 && unpackZDC_)
+      {
+	if (fed.size()==0)
+	  {
+	    edm::LogWarning("EmptyData") << "No data for FED "<< *i;
```
Comment (slava77) [inline EventFilter/CastorRawToDigi/plugins/CastorRawToDigi.cc]: please hide this and the other on L161 with an ``` if  (complainEmptyData_)  ``` as it is done for the rest of castor FEDs.
We can't accept this warning to show up in every event for most of the data that we take (most of it doesn't have ZDC).

File: RecoLocalCalo/HcalRecProducers/src/ZdcHitReconstructor.cc [hunk 3]
```diff
@@ -97,8 +98,14 @@ void ZdcHitReconstructor::produce(edm::Event& e, const edm::EventSetup& eventSet
   
    if (det_==DetId::Calo && subdet_==HcalZDCDetId::SubdetectorId) {
      edm::Handle<ZDCDigiCollection> digi;
-     e.getByToken(tok_input_,digi);
+     e.getByToken(tok_input_hcal,digi);
      
+     if(digi->size() == 0) {
+       e.getByToken(tok_input_castor,digi);
+       if(digi->size() != 0) std::cout<<"No ZDC info in hcalDigis, using castorDigis. This is normal for all Run2 datasets"<<std::endl;
```
Comment (slava77) [inline RecoLocalCalo/HcalRecProducers/src/ZdcHitReconstructor.cc]: cout is not allowed in production code.

I think this check should be done only when we really expect ZDC, which is probably just the HeavyIons scenario,
but also likely the pp setup for pPb processing.
For the former, in  Configuration/StandardSequences/python/ReconstructionHeavyIons_cff.py, and for the latter in RecoHI/Configuration/python/customise_PPwithHI.py.
You can also just change this to edm::LogInfo uniformly and then hope that someone will look at LogInfo (this severity level is not displayed/logged in production).

File: RecoLocalCalo/HcalRecProducers/src/ZdcSimpleReconstructor.cc [hunk 4]
```diff
@@ -60,7 +62,13 @@ void ZdcSimpleReconstructor::produce(edm::Event& e, const edm::EventSetup& event
   
   if (det_==DetId::Calo && subdet_==HcalZDCDetId::SubdetectorId) {
     edm::Handle<ZDCDigiCollection> digi;
-    e.getByToken(tok_input_,digi);
+    e.getByToken(tok_input_castor,digi);
+     
+     if(digi->size() == 0) {
+       e.getByToken(tok_input_hcal,digi);
+       if(digi->size() != 0) std::cout<<"No ZDC info in castorDigis, using hcalDigis. This is normal for all Run1 datasets"<<std::endl;
+       else std::cout<<"No ZDC info in either castorDigis or hcalDigis"<<std::endl;
```
Comment (slava77) [inline RecoLocalCalo/HcalRecProducers/src/ZdcSimpleReconstructor.cc]: same comments here as in 
https://github.com/cms-sw/cmssw/pull/13882/files#r58400048

**Comment by cferraio** [comment]
Comments addressed, new edits committed.

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoLocalCalo/HcalRecProducers/src/ZdcHitReconstructor.cc [hunk 5]
```diff
@@ -97,8 +98,14 @@ void ZdcHitReconstructor::produce(edm::Event& e, const edm::EventSetup& eventSet
   
    if (det_==DetId::Calo && subdet_==HcalZDCDetId::SubdetectorId) {
      edm::Handle<ZDCDigiCollection> digi;
-     e.getByToken(tok_input_,digi);
+     e.getByToken(tok_input_hcal,digi);
      
+     if(digi->size() == 0) {
+       e.getByToken(tok_input_hcal,digi);
+       if(digi->size() == 0) 
+       	 edm::LogInfo("ZdcHitReconstructor") << "No ZDC info found in either castorDigis or hcalDigis." << std::endl;
```
Comment (slava77) [inline RecoLocalCalo/HcalRecProducers/src/ZdcHitReconstructor.cc]: Where is the attempt to read from castor digis?

**Comment by slava77** [comment]
@cferraio 
It looks like both of the ZDC hit reconstructors are missing the part where they read from castor digis.

**Comment by cferraio** [comment]
hi @slava77, silly mistake on my end - the code should have read to look for hcal digis first, then castor, instead of hcal and the hcal. i've made this change:

e.getByToken(tok_input_hcal,digi);
     if(digi->size() == 0) {
       e.getByToken(tok_input_castor,digi);

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/13882	3bcbf85
- changes in the code are in line with the description and with what was integrated in 75X
- jenkins tests pass and comparisons with baseline show expected changes: addition of ZDCDataFramesSorted_castorDigis collection; also in HcalUnpackerReport_castorDigis unmappedDigis for run1 and MC tests  there is a large increase in the collection size corresponding to pseudo-errors, e.g. in Run1 HI data in wf 140.53 the HcalUnpackerReport size increases by about 1kB  (a cleaner solution telling the unpacker not to look/complain for these may help)
- tested locally in CMSSW_8_1_X_2016-04-04-1100 with Run2 HI data HIHardProbes2015 run 262694 LS100: 
    - ZDCDataFramesSorted_castorDigis appear in the outputs as expected and there is a small increase in the size of HcalUnpackerReport of castorDigis  (70 bytes)
    - with the latest update the zdc hits now have some non-default values (black "ref" is actually with this PR and red "New" is the baseline). 
all_orig-testhirun2vssign683-testhirun2_runhi2011wf140p53c_zdcrechitssorted_zdcreco__rereco_obj_obj_time
times are apparently not calibrated or the hits are some kind of noise.
all_orig-testhirun2vssign683-testhirun2_runhi2011wf140p53c_log10zdcrechitssorted_zdcreco__rereco_obj_obj_energy
In this plot log10(0) shows up at 0 (instead of - large unmber)

**Comment by slava77** [comment]
@cferraio 
(it may be too early to be sure about 80X to be used for pPb); you may want to make a backport to 80X soon.

**Comment by davidlange6** [comment]
@cferraio - could you rebase this PR? Thanks - it no longer merges with 81x for some reason.

**Comment by cferraio** [comment]
hi @davidlange6, what exactly does that entail? still fairly new to the whole process. thanks!

**Comment by slava77** [comment]
in your old test release area
```
git fetch official-cmssw CMSSW_8_1_X_2016-04-13-1100:CMSSW_8_1_X_2016-04-13-1100
git checkout -b ZdcUnpacker_DigiRecoFix81X-rebased
git rebase CMSSW_8_1_X_2016-04-13-1100
# once everything works, push this to the repository
git push --force my-cmssw ZdcUnpacker_DigiRecoFix81X-rebased:ZdcUnpacker_DigiRecoFix81X
```

if the rebase is trivial, re-testing in the latest release is not required.
Still, it would help if you check the rebased version in the new release.

**Comment by cferraio** [comment]
hi @slava77, looks like my "old test release area" is too old, CMSSW_8_1_X_2016-03-30-2300 doesn't seem to be supported any longer. what do you recommend?

**Comment by slava77** [comment]
On 4/14/16 9:29 AM, Chris Ferraioli wrote:
> hi @slava77 <https://github.com/slava77>, looks like my "old test
> release area" is too old, CMSSW_8_1_X_2016-03-30-2300 doesn't seem to be
> supported any longer. what do you recommend?
> 

you can follow the instructions using cmsenv setup from an available
release,
just cd to the old area and execute git instructions there.
Let me know if you find problems with that.

**Comment by cferraio** [comment]
@slava77 looks like I have some apparent git issue, the rebase fails:

```
[cferraio@lxplus0056 src]$ git rebase CMSSW_8_1_X_2016-04-13-1100
warning: refname 'CMSSW_8_1_X_2016-04-13-1100' is ambiguous.
warning: refname 'CMSSW_8_1_X_2016-04-13-1100' is ambiguous.
First, rewinding head to replay your work on top of it...
fatal: unable to read source tree (2bf8aedb0285b44b1036716a5d0c3c41c2c1026a)

git encountered an error while preparing the patches to replay
these revisions:

    cd8f381fe03e5ba30a009a8da24cde5ac52616db..7edb6620c2ee69397df4d6d7a5a7cdacbfdbe03d

As a result, git cannot rebase them.
```

**Comment by slava77** [comment]
OK
then I suppose you can do about the same in the new area ("latestRelease"):
```
git fetch my-cmssw ZdcUnpacker_DigiRecoFix81X:ZdcUnpacker_DigiRecoFix81X
git cms-addpkg PackageName (do this for every package that's modified in ZdcUnpacker_DigiRecoFix81X)
git checkout ZdcUnpacker_DigiRecoFix81X
git rebase latestRelease (put the appropriate value here)
# after everything is done
git push --force my-cmssw ZdcUnpacker_DigiRecoFix81X
```

**Comment by cferraio** [comment]
I'm still running in to issues, I'm not able to do the git fetch. 

```
[cferraio@lxplus005 private]$ cmsrel CMSSW_8_1_X_2016-04-14-1100
WARNING: Developer's area is created for non-production architecture slc6_amd64_gcc493. Production architecture for this release is slc6_amd64_gcc530.
[cferraio@lxplus005 private]$ cd CMSSW_8_1_X_2016-04-14-1100/src/
[cferraio@lxplus005 src]$ cmsenv
[cferraio@lxplus005 src]$ git fetch my-cmssw ZdcUnpacker_DigiRecoFix81X:ZdcUnpacker_DigiRecoFix81X
fatal: Refusing to fetch into current branch refs/heads/ZdcUnpacker_DigiRecoFix81X of non-bare repository
Unexpected end of command stream
```

And if I try to add packages, nothing happens.
```
[cferraio@lxplus005 src]$ git cms-addpkg EventFilter/CastorRawToDigi
No release tags specified, using default CMSSW_8_1_X_2016-04-14-1100.
[cferraio@lxplus005 src]$ ls
total 0
```

**Comment by slava77** [comment]
is .git already in your CMSSW_8_1_X_2016-04-14-1100/src?
run "git cms-init" before any other commands

**Comment by cferraio** [comment]
Just tried again. Started from a fresh release (deleted then cmsrel), added git cms-init after a cmsenv, ran into the same fetch issue.

**Comment by cferraio** [comment]
There was no .git either before or after "git cms-init", but it appears if I just do "git init". Then when trying the fetch command, i get:

```
[cferraio@lxplus003 src]$ git fetch my-cmssw ZdcUnpacker_DigiRecoFix81X:ZdcUnpacker_DigiRecoFix81X
fatal: 'my-cmssw' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.

```

Perhaps there's an issue with my git repo?

**Comment by slava77** [comment]
what's the result of "git cms-init"
does it print anything?

what about calling some "git cms-addpkg EventFilter/CastorRawToDigi"
do you get anything?

something is wrong with your setup.

maybe go back to basics as in 
http://cms-sw.github.io/faq.html#working-with-cmssw-on-github
and see if that works?

**Comment by slava77** [comment]
I followed the instructions in my area and it worked (modulo the part of getting code from your repo and pushing it over to mine).

```
cd CMSSW_8_1_X_2016-04-14-1100/src
eval `scramv1 ru -sh`
git cms-init
git remote add cferraio-cmssw https://github.com/cferraio/cmssw.git
git remote add cferraio-cmssw https://github.com/cferraio/cmssw.git
git fetch cferraio-cmssw ZdcUnpacker_DigiRecoFix81X:ZdcUnpacker_DigiRecoFix81X
git cms-addpkg Configuration/StandardSequences
git cms-addpkg EventFilter/CastorRawToDigi
git cms-addpkg RecoLocalCalo/Configuration
git cms-addpkg RecoLocalCalo/HcalRecProducers
git checkout ZdcUnpacker_DigiRecoFix81X
git rebase CMSSW_8_1_X_2016-04-14-1100
# fix conflict, 
# git add git add RecoLocalCalo/Configuration/python/RecoLocalCalo_EventContent_cff.py; 
# git rebase --continue
git push my-cmssw ZdcUnpacker_DigiRecoFix81X:cferraio-ZdcUnpacker_DigiRecoFix81X
```

Since this PR was ~done and you appear to be stuck,
I could make a PR with my rebased version
https://github.com/cms-sw/cmssw/compare/CMSSW_8_1_X...slava77:cferraio-ZdcUnpacker_DigiRecoFix81X

**Comment by cferraio** [comment]
Ah, good, I'm glad it works. I had just typed up a message about how I had some sort of issue with what branch I'm working on. Even using the commands you just sent I still run into errors. If it isn't too much of a hassle, it would be great if you could finalize the rebased version. Thanks a lot for the help. 

I still have to figure out what went wrong, It's affecting checking out packages, so something is still off with my git setup.

**Comment by slava77** [comment]
-1
merge conflicts due to context line changes from  #13948 merged in CMSSW_8_1_X_2016-04-13-1100
superseded by #14082

**Pull request closed without merging**